diff --git a/Lib/uuid.py b/Lib/uuid.py
index e863b63187..116570ac4a 100644
--- a/Lib/uuid.py
+++ b/Lib/uuid.py
@@ -677,6 +677,8 @@ def uuid1(node=None, clock_seq=None):
     address.  If 'clock_seq' is given, it is used as the sequence number;
     otherwise a random 14-bit sequence number is chosen."""
 
+    raise RuntimeError('uuid.uuid1() is a bad idea')
+
     # When the system provides a version-1 UUID generator, use it (but don't
     # use UuidCreate here because its UUIDs don't conform to RFC 4122).
     if _generate_time_safe is not None and node is clock_seq is None:
